umh-core/pkg/service/filesystem/buffered_filesystem.go

```diff
@@ -0,0 +1,786 @@
+package filesystem
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/united-manufacturing-hub/united-manufacturing-hub/umh-core/pkg/logger"
+)
+
+// CachedFile represents a file or directory in the filesystem
+type CachedFile struct {
+	Info  os.FileInfo
+	IsDir bool
+}
+
+// DirectoryCache represents a snapshot of a directory tree
+type DirectoryCache struct {
+	Files map[string]*CachedFile // relative path -> file info
+}
+
+// ReadDirectoryTree reads a directory tree from disk and returns a cache of its contents
+func ReadDirectoryTree(ctx context.Context, service Service, root string) (*DirectoryCache, error) {
+
+	dc := &DirectoryCache{
+		Files: make(map[string]*CachedFile),
+	}
+
+	// First check if root exists
+	rootInfo, err := service.Stat(ctx, root)
+	if err != nil {
+		return nil, fmt.Errorf("failed to walk directory tree: %w", err)
+	}
+	if !rootInfo.IsDir() {
+		return nil, fmt.Errorf("failed to walk directory tree: root is not a directory")
+	}
+
+	// Start with root directory
+	entries, err := service.ReadDir(ctx, root)
+	if err != nil {
+		return nil, fmt.Errorf("failed to walk directory tree: %w", err)
+	}
+
+	// Process each entry recursively
+	for _, entry := range entries {
+		fullPath := filepath.Join(root, entry.Name())
+		relPath, err := filepath.Rel(root, fullPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get relative path: %w", err)
+		}
+
+		info, err := entry.Info()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get file info: %w", err)
+		}
+
+		// Skip root directory
+		if relPath == "." {
+			continue
+		}
+
+		isDir := info.IsDir()
+
+		// Check if this is a symlink
+		if info.Mode()&os.ModeSymlink != 0 {
```

Comment:
What happens if the symlink links to an already read directory (e.g. loop)

Reply:
I already thought about, but decided not to do a loop prevention as there are no external influences in here that coudl cause this